### PR TITLE
SkyWalking CDS implementation

### LINKED ISSUE #53: Skywalking CDS Support
Related with https://github.com/apache/skywalking/issues/6114.
Add support for Skywalking CDS that agent receives and updates config dynamically sent from OAP.

### PR DESCRIPTION
**Pull request by Shikugawa**
Close #53

### CODE REVIEW

File: example/sample.cc [hunk 1]
```diff
@@ -42,8 +42,9 @@ int main() {
     std::string context = req.get_header_value(kPropagationHeader.data());
 
     TracingContextPtr tracing_context;
+
     if (!context.empty()) {
-      // 2. Create tracing context with propagated information.
+      // 2. Create segment context with propagated information.
```
Comment (wu-sheng) [inline example/sample.cc:47]: I think this is a misoperation? I think `create tracing context` is correct from word reading.

**Comment by wu-sheng** [comment]
2 questions
1. (Not related to CDS directly) Which service name is being used when cpp2sky in Envoy?
2. How will CDS work as in Envoy, we use the Envoy gRPC implementation? Will we still use that client? With a daemon thread working?

**Comment by Shikugawa** [comment]
> 2 questions
> 
> 1. (Not related to CDS directly) Which service name is being used when cpp2sky in Envoy?
> 2. How will CDS work as in Envoy, we use the Envoy gRPC implementation? Will we still use that client? With a daemon thread working?

1. service name is configured on Envoy from envoy config.
2. Maybe these types of configuration can be configured via Envoy LDS. But it should be conducted from Skywalking then this can be needed. It can be introduced on main thread and periodic request like healthcheck agent on Envoy. And It should be managed with Envoy gRPC implementation because the gRPC abstraction is needed.

**Comment by wu-sheng** [comment]
> Maybe these types of configuration can be configured via Envoy LDS. But it should be conducted from Skywalking then this can be needed

LDS mostly, should rely on Envoy's hot reload, right?

**Comment by Shikugawa** [comment]
> LDS mostly, should rely on Envoy's hot reload, right?

No. LDS will update config without reloading Envoy process.

**Comment by wu-sheng** [comment]
> > LDS mostly, should rely on Envoy's hot reload, right?
> 
> No. LDS will update config without reloading Envoy process.

OK, then, you could choose to support both. LDS is from Istio control, and CDS is from SkyWalking's. SkyWalking could provide consistent control with all language agents in the app, and LDS could work with Istio other settings.

**Comment by Shikugawa** [comment]
@wu-sheng PTAL

File: README.md [hunk 2]
```diff
@@ -122,6 +125,23 @@ tracing_span->endSpan();
 tracer->report(std::move(tracing_context));
 ```
 
+#### Skywalking CDS
```
Comment (wu-sheng) [inline README.md:128]: After this PR merged or released, I think you need to reword the `Value Format Example` of `configuration-discovery.default.agentConfigurations` here, https://github.com/apache/skywalking/blob/master/docs/en/setup/backend/dynamic-config.md

Change to something like, 
```
[Java agent](../service-agent/java-agent/configuration-discovery.md), [cpp2sky SDK](your link)
```

**Review by wu-sheng — APPROVED** [review]

**Pull request merged**